### PR TITLE
Votes now show the percentage of total votes each choice got | Adds weighted random vote type | Map votes are now weighted random

### LINKED ISSUE #80408: Map vote log is split into several lines which makes it completely useless in parsed-logs which expects things that start with timestamps
image

The best fix for this is getting map votes to be properly logged, ideally either through JSON logs or their own log files, because I really do need this information

### PR DESCRIPTION
**Pull request by ZephyrTFA**
## About The Pull Request

See the title.
## Why It's Good For The Game

Mothblocks wanted map votes to not be simple winner take all and wanted it to be randomly chosen based on votes.
image

## Changelog
:cl:
add: Map Votes are now weighted random.
add: Custom Votes can now take advantage of Weighted Random winner selection
del: Removed Herobrine from the game
/:cl:

### CODE REVIEW

File: code/__DEFINES/maths.dm [hunk 1]
```diff
@@ -43,6 +43,8 @@
 
 #define ROUND_UP(x) ( -round(-(x)))
 
+#define DIGITS(x) (ROUND_UP(log(10, x)))
```
Comment (Mothblocks) [inline code/__DEFINES/maths.dm:46]: doc

File: code/datums/votes/_vote_datum.dm [hunk 2]
```diff
@@ -123,33 +125,43 @@
  */
 /datum/vote/proc/get_vote_result(list/non_voters)
 	RETURN_TYPE(/list)
+	SHOULD_CALL_PARENT(TRUE)
+
+	switch(winner_method)
+		if(VOTE_WINNER_METHOD_NONE)
+			return list()
+
```
Comment (Mothblocks) [inline code/datums/votes/_vote_datum.dm:133]: Nit: don't like the whitespace in between here

File: code/datums/votes/_vote_datum.dm [hunk 3]
```diff
@@ -161,17 +173,48 @@
...
 	if(override_question)
 		returned_text += span_bold(override_question)
 	else
 		returned_text += span_bold("[capitalize(name)] Vote")
 
+	returned_text += "\nWinner Selection: "
+	switch(winner_method)
+		if(VOTE_WINNER_METHOD_NONE)
+			returned_text += "None"
+
```
Comment (Mothblocks) [inline code/datums/votes/_vote_datum.dm:186]: same

File: code/datums/votes/_vote_datum.dm [hunk 4]
```diff
@@ -161,17 +173,48 @@
...
+	if(total_votes <= 0)
+		return span_bold("Vote Result: Inconclusive - No Votes!")
+
+	returned_text += "\nResults:"
 	for(var/option in choices)
-		returned_text += "\n[span_bold(option)]: [choices[option]]"
+		returned_text += "\n"
+		var/votes = choices[option]
+		var/percentage_text = ""
+		if(votes > 0)
```
Comment (Mothblocks) [inline code/datums/votes/_vote_datum.dm:205]: Is this necessary? `0 / ` etc is just zero. Seems like a waste of code

File: code/__DEFINES/subsystems.dm [hunk 5]
```diff
@@ -334,3 +334,10 @@
 #define VOTE_COUNT_METHOD_SINGLE 1
 /// Approval voting. Any number of selections per person, and the selection with the most votes wins.
 #define VOTE_COUNT_METHOD_MULTI 2
+
+/// The choice with the most votes wins. Ties are broken by the first choice to reach that number of votes.
+#define VOTE_WINNER_METHOD_SIMPLE 1
+/// The winning choice is selected randomly based on the number of votes each choice has.
+#define VOTE_WINNER_METHOD_RANDOM 2
```
Comment (Mothblocks) [inline code/__DEFINES/subsystems.dm:341]: Nit: Would be better as "METHOD_WEIGHTED_RANDOM", more obvious

File: code/__DEFINES/subsystems.dm [hunk 6]
```diff
@@ -334,3 +334,10 @@
 #define VOTE_COUNT_METHOD_SINGLE 1
 /// Approval voting. Any number of selections per person, and the selection with the most votes wins.
 #define VOTE_COUNT_METHOD_MULTI 2
+
+/// The choice with the most votes wins. Ties are broken by the first choice to reach that number of votes.
+#define VOTE_WINNER_METHOD_SIMPLE 1
```
Comment (Mothblocks) [inline code/__DEFINES/subsystems.dm:339]: These should be strings unless there's a reason for them to be numbers--it's harder to VV these

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Nice

File: code/datums/votes/_vote_datum.dm (hunk 4 above)
Comment (ZephyrTFA) [reply code/datums/votes/_vote_datum.dm:205]: log of 0 is a runtime error

File: code/datums/votes/custom_vote.dm [hunk 7]
```diff
@@ -29,6 +29,27 @@
...
+		vote_creator,
+		"How should the vote winner be determined?",
+		"Winner Method",
+		list("Simple", "Weighted Random", "No Winner"),
+		default = "Simple",
+	)
+	switch(custom_win_method)
+		if("Simple")
+			winner_method = VOTE_WINNER_METHOD_SIMPLE
+
```
Comment (Mothblocks) [inline code/datums/votes/custom_vote.dm:42]: same nit here on whitespace

File: code/controllers/subsystem/vote.dm [hunk 8]
```diff
@@ -86,11 +86,31 @@ SUBSYSTEM_DEF(vote)
...
-	log_vote(to_display)
+	var/total_votes = 0
+	var/list/vote_choice_data = list()
+	for(var/choice in current_vote.choices)
+		var/choice_votes = current_vote.choices[choice]
+		vote_choice_data["[choice]"] = choice_votes
+
+	// stringify the winners to prevent potential unimplemented serialization errors.
+	// Perhaps this can be removed in the future and we assert that vote choices must implement serialization.
+	var/final_winner_string = !isnull(final_winner) ? "[final_winner]" : null
```
Comment (Mothblocks) [inline code/controllers/subsystem/vote.dm:97]:
```suggestion
	var/final_winner_string = final_winner && "[final_winner]"
```

**Comment by Mothblocks** [comment]
Log entry:

```
{"ts":"2023-12-17 11:47:18.138","cat":"game-vote","msg":"<span class='bold'>Map Vote</span>\\nWinner Selection: Weighted Random\\nResults:\\n 33.3% | <span class='bold'>Ice Box Station</span>: 5\\n  6.7% | <span class='bold'>Tramstation</span>: 1\\n  6.7% | <span class='bold'>Birdshot Station</span>: 1\\n 53.3% | <span class='bold'>MetaStation</span>: 8\\n<span class='bold'>\\nVote Result: MetaStation</span>","data":{"choices":{"Ice Box Station":5,"Tramstation":1,"Birdshot Station":1,"MetaStation":8},"total":0,"winners":{"MetaStation":null},"final_winner":"MetaStation"},"w-state":{"tick_usage":17.9136,"tick_lag":0.5,"time":51100,"timestamp":"1702813638.139092"},"s-store":[],"id":51364,"s-ver":"1.0.0"}
```

Note the `winners` being a keyed dictionary with value null? What's up with that?

**Comment by ZephyrTFA** [comment]
> Log entry:
> 
> ```
> {"ts":"2023-12-17 11:47:18.138","cat":"game-vote","msg":"<span class='bold'>Map Vote</span>\\nWinner Selection: Weighted Random\\nResults:\\n 33.3% | <span class='bold'>Ice Box Station</span>: 5\\n  6.7% | <span class='bold'>Tramstation</span>: 1\\n  6.7% | <span class='bold'>Birdshot Station</span>: 1\\n 53.3% | <span class='bold'>MetaStation</span>: 8\\n<span class='bold'>\\nVote Result: MetaStation</span>","data":{"choices":{"Ice Box Station":5,"Tramstation":1,"Birdshot Station":1,"MetaStation":8},"total":0,"winners":{"MetaStation":null},"final_winner":"MetaStation"},"w-state":{"tick_usage":17.9136,"tick_lag":0.5,"time":51100,"timestamp":"1702813638.139092"},"s-store":[],"id":51364,"s-ver":"1.0.0"}
> ```
> 
> Note the `winners` being a keyed dictionary with value null? What's up with that?

This is because all lists are assumed to be associative for logging purposes. Once we get actual parity for dictionaries and lists this can be changed.

**Comment by JohnFulpWillard** [comment]
YOU REMOVED WHAT !?

**Comment by Sakamoto4ka** [comment]
Code_VaJBeAJij8
herobrine still in the game

**Pull request merged**